awx/main/tasks/jobs.py

```diff
@@ -444,164 +442,8 @@ def final_run_hook(self, instance, status, private_data_dir, fact_modification_t
-
-        elif self.recent_event_timings.maxlen:
-            self.recent_event_timings.append(time.time())
-
-        event_data.setdefault(self.event_data_key, self.instance.id)
-        self.dispatcher.dispatch(event_data)
-        self.event_ct += 1
-
-        '''
-        Handle artifacts
-        '''
-        if event_data.get('event_data', {}).get('artifact_data', {}):
-            self.instance.artifacts = event_data['event_data']['artifact_data']
-            self.instance.save(update_fields=['artifacts'])
-
-        return False
-
-    def cancel_callback(self):
-        """
-        Ansible runner callback to tell the job when/if it is canceled
-        """
-        unified_job_id = self.instance.pk
-        self.instance = self.update_model(unified_job_id)
-        if not self.instance:
-            logger.error('unified job {} was deleted while running, canceling'.format(unified_job_id))
-            return True
-        if self.instance.cancel_flag or self.instance.status == 'canceled':
-            cancel_wait = (now() - self.instance.modified).seconds if self.instance.modified else 0
-            if cancel_wait > 5:
-                logger.warn('Request to cancel {} took {} seconds to complete.'.format(self.instance.log_format, cancel_wait))
-            return True
-        return False
-
-    def finished_callback(self, runner_obj):
-        """
-        Ansible runner callback triggered on finished run
-        """
-        event_data = {
-            'event': 'EOF',
-            'final_counter': self.event_ct,
-            'guid': self.guid,
-        }
-        event_data.setdefault(self.event_data_key, self.instance.id)
-        self.dispatcher.dispatch(event_data)
-
-    def status_handler(self, status_data, runner_config):
-        """
-        Ansible runner callback triggered on status transition
-        """
-        if status_data['status'] == 'starting':
-            job_env = dict(runner_config.env)
-            '''
-            Take the safe environment variables and overwrite
-            '''
-            for k, v in self.safe_env.items():
-                if k in job_env:
-                    job_env[k] = v
-            from awx.main.signals import disable_activity_stream  # Circular import
-
-            with disable_activity_stream():
-                self.instance = self.update_model(self.instance.pk, job_args=json.dumps(runner_config.command), job_cwd=runner_config.cwd, job_env=job_env)
-        elif status_data['status'] == 'error':
-            result_traceback = status_data.get('result_traceback', None)
-            if result_traceback:
-                from awx.main.signals import disable_activity_stream  # Circular import
-
-                with disable_activity_stream():
-                    self.instance = self.update_model(self.instance.pk, result_traceback=result_traceback)
+    def attach_callback(self):
+        self.runner_callback = RunnerCallback()
```

Comment:
The two classes ```BaseTak``` and ```RunnerCallback``` are coupled here.

Reply:
If we look at this coupling we can easily do something like this :

--- change a value inside of a callback instance
example : self.runner_callback.some_variable = new_value

What I am stuck at is the following:
--- if something  changes inside the runner.callback ( which is inside ```RunnerCallback object```, how would that change be propagated to  outside class , in this case ```BaseTask``` . 

I can go ahead and implement something like this : https://stackoverflow.com/questions/2024566/how-to-access-outer-class-from-an-inner-class , but it seems to overcomplicate things.

Reply:
I have removed the ```attach callback()``` function